javalib/src/main/scala/java/math/Multiplication.scala

```diff
@@ -0,0 +1,454 @@
+
+  /**
+   * Multiplies two BigIntegers. Implements traditional scholar algorithm
+   * described by Knuth.
+   *
+   * <br>
+   * <tt>
+   *         <table border="0">
+   * <tbody>
+   *
+   *
+   * <tr>
+   * <td align="center">A=</td>
+   * <td>a<sub>3</sub></td>
+   * <td>a<sub>2</sub></td>
+   * <td>a<sub>1</sub></td>
+   * <td>a<sub>0</sub></td>
+   * <td></td>
+   * <td></td>
+   * </tr>
+   *
+   * <tr>
+   * <td align="center">B=</td>
+   * <td></td>
+   * <td>b<sub>2</sub></td>
+   * <td>b<sub>1</sub></td>
+   * <td>b<sub>1</sub></td>
+   * <td></td>
+   * <td></td>
+   * </tr>
+   *
+   * <tr>
+   * <td></td>
+   * <td></td>
+   * <td></td>
+   * <td>b<sub>0</sub>*a<sub>3</sub></td>
+   * <td>b<sub>0</sub>*a<sub>2</sub></td>
+   * <td>b<sub>0</sub>*a<sub>1</sub></td>
+   * <td>b<sub>0</sub>*a<sub>0</sub></td>
+   * </tr>
+   *
+   * <tr>
+   * <td></td>
+   * <td></td>
+   * <td>b<sub>1</sub>*a<sub>3</sub></td>
+   * <td>b<sub>1</sub>*a<sub>2</sub></td>
+   * <td>b<sub>1</sub>*a1</td>
+   * <td>b<sub>1</sub>*a0</td>
+   * </tr>
+   *
+   * <tr>
+   * <td>+</td>
+   * <td>b<sub>2</sub>*a<sub>3</sub></td>
+   * <td>b<sub>2</sub>*a<sub>2</sub></td>
+   * <td>b<sub>2</sub>*a<sub>1</sub></td>
+   * <td>b<sub>2</sub>*a<sub>0</sub></td>
+   * </tr>
+   *
+   * <tr>
+   * <td></td>
+   * <td>______</td>
+   * <td>______</td>
+   * <td>______</td>
+   * <td>______</td>
+   * <td>______</td>
+   * <td>______</td>
+   * </tr>
+   *
+   * <tr>
+   *
+   * <td align="center">A*B=R=</td>
+   * <td align="center">r<sub>5</sub></td>
+   * <td align="center">r<sub>4</sub></td>
+   * <td align="center">r<sub>3</sub></td>
+   * <td align="center">r<sub>2</sub></td>
+   * <td align="center">r<sub>1</sub></td>
+   * <td align="center">r<sub>0</sub></td>
+   * <td></td>
+   * </tr>
+   *
+   * </tbody>
+   * </table>
+   *
+   *</tt>
+   *
+   * @param op1 first factor of the multiplication {@code op1 >= 0}
+   * @param op2 second factor of the multiplication {@code op2 >= 0}
+   * @return a {@code BigInteger} of value {@code op1 * op2}
+   */
+  def multiplyPAP(a: BigInteger, b: BigInteger): BigInteger = {
+    val aLen = a.numberLength
+    val bLen = b.numberLength
+    val resLength = aLen + bLen
+    val resSign =
+      if (a.sign != b.sign) -1
+      else 1
+
+    if (resLength == 2) {
+      val v = unsignedMultAddAdd(a.digits(0), b.digits(0), 0, 0)
+      val valueLo = v.toInt
+      val valueHi = (v >>> 32).toInt
+      if (valueHi == 0) new BigInteger(resSign, valueLo)
+      else new BigInteger(resSign, 2, Array(valueLo, valueHi))
+    } else {
+      val aDigits = a.digits
+      val bDigits = b.digits
+      val resDigits = new Array[Int](resLength)
+      multArraysPAP(aDigits, aLen, bDigits, bLen, resDigits)
+      val result = new BigInteger(resSign, resLength, resDigits)
+      result.cutOffLeadingZeroes()
+      result
+    }
+  }
+
+  def pow(base: BigInteger, exponent: Int): BigInteger = {
+    var _exponent = exponent
```

Comment:
Not a big fan of the `_` here. Consider renaming this as `exp`. Or, possibly better, writing the loop as an inline tailrec function.